### PR TITLE
[RLlib] Algorithm `step()` fixes: evaluation should NOT be part of timed `training_step` loop.

### PR DESCRIPTION
**Pull request by sven1977**
Algorithm `step()` fixes: evaluation should NOT be part of timed `training_step` loop.
* The evaluation step used to happen within the min_time_s_per_iteration-timed `training_step()` loop, which causes performance issues when `training_step()` runs fast (e.g. in async sampling/training algos, like IMPALA/APEX).
* Added `remove_in_flight_requests` option to `AsyncRequestsManager::remove_workers`, such that in case of errors, a faulty workers' requests (in-flight) can be removed entirely. Not doing so would cause these pending requests to keep failing over and over again leading to a stuck `training_step()` loop.

Basically before the fix, we did:
```
def step():
  while not min_time_per_iteration reached:
    self.training_step()  # formerly known as `training_iteration()`
    self.evaluate()
```
^ this would become a problem if the training_iteration() method runs very fast (like it does in the async algos: APPO/IMPALA). Then evaluation would be called all the time, slowing things down.

Fix: It should be (in the sequential case; no auto-duration):
```
def step():
  while not min_time_per_iteration reached:
    self.training_step()  # formerly known as `training_iteration()`
  self.evaluate()
```
and in the parallel case (with auto-duration):
```
def step():
  # -- run in a thread --
  while not min_time_per_iteration reached:
    self.training_step()  # formerly known as `training_iteration()`
  # -- end: run in a thread --

  while above thread not done:
    self.evaluate()
```

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by avnishn — COMMENTED** [review]
1 Quick question:

Did we stop using step_ctx.should_stop in favor of evaluate this itr? 

If so this change makes sense to me in the context.

**Comment by sven1977** [comment]
We are mostly fixing a logical error in our current `train()` sequence. What we want to have is this (for the plain non-parallel case):

```
def train():  # implemented by tune.Trainable superclass
   ...
   self.step()
   ...


def step():  # implemented by Algorithm   
   while min_time_s_per_iteration not reached:
        results = training_step()

    results.update(self.evaluate())

    return results
```

Before this PR, the `self.evaluate` was part of the while-loop, with the consequence that for very quick `training_step()` execution times (e.g. for IMPALA/APPO), we would evaluate all the time, which would slow down things a lot.

File: doc/source/rllib/rllib-training.rst [hunk 1]
```diff
@@ -148,12 +148,16 @@ Here are some rules of thumb for scaling training with RLlib.
 
 
 In case you are using lots of workers (``num_workers >> 10``) and you observe worker failures for whatever reasons, which normally interrupt your RLlib training runs, consider using
-the config settings ``ignore_worker_failures=True`` or ``recreate_failed_workers=True``:
+the config settings ``ignore_worker_failures=True``, ``recreate_failed_workers=True``, or ``restart_failed_sub_environments=True``:
 
-``ignore_worker_failures=True`` allows your Algorithm to not crash due to a single worker error, but to continue for as long as there is at least one functional worker remaining.
-``recreate_failed_workers=True`` will have your Algorithm attempt to replace/recreate any failed worker(s) with a new one.
+``ignore_worker_failures``: When set to True, your Algorithm will not crash due to a single worker error, but to continue for as long as there is at least one functional worker remaining.
```
Comment (ArturNiederfahrenhorst) [inline doc/source/rllib/rllib-training.rst:153]: ... ,but continue for ...
or ... . Instead it will continue for ....

File: rllib/execution/parallel_requests.py [hunk 2]
```diff
@@ -371,15 +371,18 @@ def add_workers(self, new_workers: Union[List[ActorHandle], ActorHandle]) -> Non
             if new_worker not in self._all_workers:
                 self._all_workers.append(new_worker)
 
-    def remove_workers(self, workers: Union[List[ActorHandle], ActorHandle]) -> None:
+    def remove_workers(
+        self,
+        workers: Union[List[ActorHandle], ActorHandle],
+        remove_in_flight_requests: bool = False,
```
Comment (ArturNiederfahrenhorst) [inline rllib/execution/parallel_requests.py:377]: Just to learn about how we use this: In what case would we want to remove a worker without removing in-flight requests?

File: rllib/algorithms/algorithm.py [hunk 3]
```diff
@@ -581,107 +576,50 @@ def step(self) -> ResultDict:
...
-                        raise e
-                # Any other exception.
-                except Exception as e:
-                    # Allow logs messages to propagate.
-                    time.sleep(0.5)
-                    raise e
+        # - No evaluation necessary, just run the next training iteration.
+        # - We have to evaluate in this training iteration, but no parallelism ->
+        #   evaluate after the training iteration is entirely done.
+        if not evaluate_this_iter or not self.config["evaluation_parallel_to_training"]:
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/algorithm.py:582]: While we're at it I think we could get of the double _if evaluate_this_iter and not self.config["evaluation_parallel_to_training"]_ in L594/L582 ?

File: rllib/algorithms/algorithm.py [hunk 4]
```diff
@@ -581,107 +576,50 @@ def step(self) -> ResultDict:
...
+            results, train_iter_ctx = self._run_one_training_iteration()
+        # Kick off evaluation-loop (and parallel train() call,
+        # if requested).
+        # Parallel eval + training.
+        else:
+            (
+                results,
+                train_iter_ctx,
+            ) = self._run_one_training_iteration_and_evaluation_in_parallel()
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/algorithm.py:591]: How would we handle a metric that utilizes both, training and evaluation, results?

File: rllib/algorithms/algorithm.py [hunk 5]
```diff
@@ -1690,14 +1619,6 @@ def _sync_weights_to_workers(
         weights = ray.put(self.workers.local_worker().save())
         worker_set.foreach_worker(lambda w: w.restore(ray.get(weights)))
 
-    def _exec_plan_or_training_step_fn(self):
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/algorithm.py:1693]: Awesome!
This means that there's "kind of" no revert to an exec plan, scary!

File: rllib/algorithms/algorithm.py [hunk 6]
```diff
@@ -2129,43 +2050,59 @@ def validate_env(env: EnvType, env_context: EnvContext) -> None:
...
+                    "To try to continue without failed "
+                    "worker(s), set `ignore_worker_failures=True`. "
+                    "To try to recover the failed worker(s), set "
+                    "`recreate_failed_workers=True`."
+                )
+                raise error
+        # Any other exception.
+        else:
+            # Allow logs messages to propagate.
+            time.sleep(0.5)
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/algorithm.py:2083]: I'm curious again. Every RayError is a RolloutWorker failure?

File: rllib/BUILD [hunk 7]
```diff
@@ -809,7 +809,7 @@ py_test(
 py_test(
     name = "test_cql",
     tags = ["team:rllib", "algorithms_dir"],
-    size = "medium",
+    size = "large",
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/BUILD:812]: We don't test learning capabilities in this test, just "compilation" + a few iterations. So no.

File: rllib/execution/parallel_requests.py (hunk 2 above)
Comment (sven1977) [reply rllib/execution/parallel_requests.py:377]: Maybe if the worker got "reassigned" to some other worker pool?
Not sure either. So far, we only use this method for when workers failed, so it makes sense here to also erase their in-flight requests tracking (otherwise, these in-flight requests would fail repeatedly and could lead to getting stuck in an endless loop).

File: rllib/algorithms/algorithm.py (hunk 3 above)
Comment (sven1977) [reply rllib/algorithms/algorithm.py:582]: ? 
You mean we should make the if check a positive one? Like
```
if evaluate_this_iter and self.config["evaluation_parallel_to_training"]:
  ...
else
  ...
```

File: rllib/algorithms/algorithm.py [hunk 8]
```diff
@@ -2303,8 +2240,112 @@ def _kwargs_for_execution_plan(self):
...
+        """
+        with concurrent.futures.ThreadPoolExecutor() as executor:
+            train_future = executor.submit(lambda: self._run_one_training_iteration())
+            # Pass the train_future into `self._run_one_evaluation()` to allow it
+            # to run exactly as long as the training iteration takes in case
+            # evaluation_duration=auto.
+            results = self._run_one_evaluation(train_future)
+            # Collect the training results from the future.
+            train_results, train_iter_ctx = train_future.result()
+            results.update(train_results)
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/algorithm.py:2343]: Should have put my comment form above here: How to write a metric that includes both, training and evaluation results?

**Review by ArturNiederfahrenhorst — APPROVED** [review]
Not seeing anything wrong here, just nits and questions.

File: rllib/algorithms/algorithm.py (hunk 6 above)
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/algorithm.py:2081]: Awesome, thanks!

File: rllib/execution/parallel_requests.py (hunk 2 above)
Comment (gjoliver) [reply rllib/execution/parallel_requests.py:377]: why not make the default True if there isn't a case where we shouldn't remove the in-flight requests?
